src/mailadm/conn.py

```diff
@@ -210,31 +185,36 @@ def get_user_list(self, token=None):
             args.append(token)
         return [UserInfo(*args) for args in self._sqlconn.execute(q, args).fetchall()]
 
-    def add_email_account(self, token_info, addr=None, password=None, tries=1):
-        for i in range(tries):
-            try:
-                return self._add_addr(token_info, addr=addr, password=password)
-            except (ValueError, DBError):
-                if i + 1 >= tries:
-                    raise
-
-    def _add_addr(self, token_info, addr, password):
-        config = self.config
+    def add_email_account(self, token_info, addr=None, password=None):
+        token_info.check_exhausted()
+        if password is None:
+            password = mailadm.util.gen_password()
         if addr is None:
             rand_part = mailadm.util.get_human_readable_id()
             username = "{}{}".format(token_info.prefix, rand_part)
-            addr = "{}@{}".format(username, config.mail_domain)
+            addr = "{}@{}".format(username, self.config.mail_domain)
         else:
-            if not addr.endswith(config.mail_domain):
+            if not addr.endswith(self.config.mail_domain):
                 raise ValueError("email {!r} is not on domain {!r}".format(
-                                 addr, config.mail_domain))
+                                 addr, self.config.mail_domain))
 
-        clear_pw, hash_pw = get_doveadm_pw(password=password)
-        self.add_user(addr=addr, hash_pw=hash_pw, date=int(time.time()),
+        self.add_user(addr=addr, date=int(time.time()),
```

Comment:
maybe it's better to first check if mailcow already has that user, therebow testing with read-only methods if mailcow is reachable?  If not then we can bail out without modifying anything.  If mailcow does not have the new user address, then usually our database should also not have the user (and will otherwise throw an exception i guess).   Probably makes sense to add a "has_user(addr)" method to mailcow.py for this.